src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,71 +280,204 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
         Logger.v(TAG, "Finished decryption");
         return decrypted;
     }
+    
+    /**
+     * Get Secret Key based on API level to use in encryption. Decryption key
+     * depends on version# since user can migrate to new Android.OS
+     *
+     * @return
+     * @throws NoSuchAlgorithmException
+     */
+    public synchronized SecretKey loadSecretKeyForEncryption() throws IOException, GeneralSecurityException {
+        // Loading key only once for performance. If API is upgraded, it will
+        // restart the device anyway. It will load the correct key for new API.
+        if (mKey != null && mMacKey != null)
```

Comment:
Why this function doesn't invoke `loadSecretKeyForDecryption` and if it fails fallback to generate new keys